cocos/network/WebSocket-libwebsocket.h

```diff
@@ -0,0 +1,155 @@
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+
+"[WebSocket module] is based in part on the work of the libwebsockets  project
+(http://libwebsockets.org)"
+
+ ****************************************************************************/
+
+#pragma once
+
+#include <string>
+#include <vector>
+#include <mutex>
+#include <memory>  // for std::shared_ptr
+#include <atomic>
+#include <condition_variable>
+
+#include "platform/CCPlatformMacros.h"
+#include "platform/CCStdC.h"
+#include "network/WebSocket.h"
+
+
+struct lws;
+struct lws_protocols;
+struct lws_vhost;
+/**
+ * @addtogroup network
+ * @{
+ */
+
+NS_CC_BEGIN
+
+class EventListenerCustom;
+
+namespace network {
+
+class WebSocket;
+class WsThreadHelper;
+
+/**
+ * WebSocket is wrapper of the libwebsockets-protocol, let the develop could call the websocket easily.
+ * Please note that all public methods of WebSocket have to be invoked on Cocos Thread.
+ */
+class CC_DLL WebSocketImpl
+{
+public:
+
+    WebSocketImpl(WebSocket *);
+
+    virtual ~WebSocketImpl();
+
+    bool init(const WebSocket::Delegate& delegate,
+              const std::string& url,
+              const std::vector<std::string>* protocols = nullptr,
+              const std::string& caFilePath = "");
+
+    void send(const std::string& message);
+
+    void send(const unsigned char* binaryMsg, unsigned int len);
+
+    void close();
+
+    void closeAsync();
+
+    WebSocket::State getReadyState();
+
+    inline const std::string& getUrl() const { return _url; }
+
+    inline const std::string& getProtocol() const { return _selectedProtocol; }
+
+
+private:
+
+    // The following callback functions are invoked in websocket thread
+    void onClientOpenConnectionRequest();
+    int onSocketCallback(struct lws *wsi, int reason, void *in, ssize_t len);
+
+    int onClientWritable();
+    int onClientReceivedData(void* in, ssize_t len);
+    int onConnectionOpened();
+    int onConnectionError(void* in, ssize_t len);
+    int onConnectionClosed();
+
+    struct lws_vhost* createVhost(struct lws_protocols* protocols, int& sslConnection);
+
+private:
+
+    std::vector<char> _receivedData;
+
+    struct lws* _wsInstance;
+    struct lws_protocols* _lwsProtocols;
+    std::string _clientSupportedProtocols;
+
+    std::shared_ptr<std::atomic<bool>> _isDestroyed;
+
+    std::mutex _closeMutex;
+    std::condition_variable _closeCondition;
+    std::vector<char*> _protocolNames;
+
+    enum class CloseState
+    {
+        NONE,
+        SYNC_CLOSING,
+        SYNC_CLOSED,
+        ASYNC_CLOSING
+    };
+    
+    CloseState _closeState;
+    std::mutex  _readyStateMutex;
+    EventListenerCustom* _resetDirectorListener;
+
+    friend class WsThreadHelper;
+    friend class WebSocketCallbackWrapper;
+
+    WebSocket *_socket = nullptr;
+
+
+    WebSocket::State       _readyState =  WebSocket::State ::CONNECTING;
+    WebSocket::Delegate*   _delegate  = nullptr;
```

Comment:
`_socket`, `_readyState`, `_delegate` are initialized here and in member initializer list. The member initializer list has higher priority.